drivers/mysql/lock.go

```diff
@@ -0,0 +1,194 @@
+	if _, err = ms.conn.ExecContext(ctx, createTableIfNotExistsQuery); err != nil {
+		return nil, err
+	}
+
+	return &Mutex{
+		key:  key,
+		conn: ms.conn,
+	}, nil
+}
+
+// lock makes a single attempt to lock the mutex, returning true only if successful.
+func (m *Mutex) tryLock(ctx context.Context) (bool, error) {
+	now := time.Now()
+	query := fmt.Sprintf("INSERT INTO %s (Id, ExpireAt) VALUES ('%s', %d)", drivers.MutexTableName, m.key, now.Add(drivers.TTL).Unix())
+	if _, err := m.conn.ExecContext(ctx, query); err != nil {
+		err2 := m.releaseLock(ctx, now)
+		if err2 == nil {
+			// lock has been released due to expiration
+			return true, nil
+		}
+		return false, errors.Wrapf(err, "failed to lock mutex")
+	}
+
+	return true, nil
+}
+
+func (m *Mutex) releaseLock(ctx context.Context, t time.Time) error {
+	e, err := m.getExpireAt(ctx)
+	if err != nil {
+		return err
+	}
+
+	if t.Unix() < e {
+		return errors.New("could not release the lock")
+	}
+
+	query := fmt.Sprintf("UPDATE %s SET ExpireAt = %d WHERE Id = '%s'", drivers.MutexTableName, t.Add(drivers.TTL).Unix(), m.key)
+	_, err = m.conn.ExecContext(ctx, query)
+	if err != nil {
+		return errors.Wrap(err, "unable to set new expireat for mutex")
+	}
+	return nil
+}
+
+func (m *Mutex) getExpireAt(ctx context.Context) (int64, error) {
+	var expireAt int64
+	query := fmt.Sprintf("SELECT ExpireAt FROM %s WHERE Id = '%s'", drivers.MutexTableName, m.key)
+	err := m.conn.QueryRowContext(ctx, query).Scan(&expireAt)
+	if err != nil {
+		return -1, errors.Wrap(err, "failed to fetch mutex from db")
+	}
+
+	return expireAt, nil
+}
+
+// refreshLock rewrites the lock key value with a new expiry, returning nil only if successful.
+func (m *Mutex) refreshLock(ctx context.Context) error {
+	e, err := m.getExpireAt(ctx)
+	if err != nil {
+		return err
+	}
+
+	tmp := time.Unix(e, 0)
+	query := fmt.Sprintf("UPDATE %s SET ExpireAt = %d WHERE Id = '%s'", drivers.MutexTableName, tmp.Add(drivers.TTL).Unix(), m.key)
+	_, err = m.conn.ExecContext(ctx, query)
+	if err != nil {
+		return errors.Wrap(err, "unable to refresh expireat for mutex")
+	}
+	return nil
+}
+
+// Lock locks m. If the mutex is already locked by any other morph instance, including the current one,
+// the calling goroutine blocks until the mutex can be locked.
+func (m *Mutex) Lock() {
+	_ = m.LockWithContext(context.Background())
+}
+
+// LockWithContext locks m unless the context is canceled. If the mutex is already locked by any other
+// instance, including the current one, the calling goroutine blocks until the mutex can be locked,
+// or the context is canceled.
+//
+// The mutex is locked only if a nil error is returned.
+func (m *Mutex) LockWithContext(ctx context.Context) error {
+	var waitInterval time.Duration
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(waitInterval):
+		}
+
+		locked, err := m.tryLock(ctx)
```

Comment:
Let's rename this to `ok`. It had me confused for a second as I didn't understand whether this meant "it failed to lock because it is locked", or "it acquired a lock".